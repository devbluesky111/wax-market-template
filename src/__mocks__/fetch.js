// Should never happen ðŸ˜‡ You are missing a status code in your `on(â€¦)` calls.
const INVALID_STATUS = 599

/**
 * Call to clean up the state of this. Wrap it in `afterEach(() => {â€¦}` or like `afterEach(reset)`
 */
export const reset = () => {
    expect(waitingRequests).toHaveLength(0)
    expect(expectedCalls).toHaveLength(0)
    expect(calls.every((call) => call.valid)).toBeTruthy()
    waitingRequests = []
    expectedCalls = []
    calls = []
    matchMode = MATCH_FIFO
}

const MATCH_FIFO = 'fifo'
const MATCH_URL = 'url'
let matchMode = MATCH_FIFO

/**
 * @param {'fifo' | 'url'} mode
 */
export const setMatchMode = (mode) => (matchMode = mode)

/**
 * Debug report for `afterEach` or just somewhere in your tests
 */
export const report = () => {
    console.log({ waitingRequests, expectedCalls, calls })
}

/**
 * List of all calls made
 * @type {(Call & { valid: boolean })[]}
 */
export let calls = []

/**
 * @typedef {{match: FetchFn<boolean>, response: any, status: number}} ExpectedCall
 */

/**
 * List of all currently expected calls to be made in the future
 * @type {ExpectedCall[]}
 */
export let expectedCalls = []

/**
 * List of all currently active Requests to be flushed
 * @type {{ url: string, flush: VoidFunction }[]}
 */
export let waitingRequests = []

/**
 * @typedef {string | [number, string] | Response} MaybeResponse
 */

/**
 * @template T
 * @typedef {(input: RequestInfo, init?: RequestInit) => T} FetchFn
 */

/**
 * @typedef {Object} Call
 * @property {string} [url]
 * @property {number} [status]
 * @property {any} [body]
 * @property {MaybeResponse} response
 **/

/**
 * @param {string} [msg]
 * @returns {() => Promise<any>}
 */
const throwThis = (msg) => () => {
    throw new Error(msg ?? 'Not implemented!')
}
/**
 * @param {string} [msg]
 * @returns {() => any}
 */
const throwThisSync = (msg) => () => {
    throw new Error(msg ?? 'Not implemented!')
}

/**
 * @param {string} url
 * @param {string | any} content
 * @param {number} [statusCode]
 * @returns {Response}
 */
export const createResponse = (url, content, statusCode = 200) => ({
    arrayBuffer: throwThis(),
    blob: async () => new Blob(content),
    // @ts-ignore
    body: 'Not implemented!',
    bodyUsed: false,
    clone: throwThisSync(),
    // @ts-ignore
    formData: 'Not implemented!',
    // @ts-ignore
    headers: 'Not implemented!',
    json: () => (typeof content === 'string' ? JSON.parse(content) : content),
    ok: statusCode >= 200 && statusCode < 300,
    redirected: false,
    statusText: 'Not implemented!',
    status: statusCode,
    text: () => Promise.resolve(typeof content === 'string' ? content : JSON.stringify(content)),
    // @ts-ignore
    type: 'Not implemented!',
    url,
})

/**
 *
 * @param {Call | MaybeResponse} call
 * @param {number} [status]
 * @param {string} [url]
 * @return {{ type: 'result', result: string, status: number, url?: string } | { type: 'response', response: Response }}
 */
const responseHandler = (call, status = 200, url = undefined) => {
    if (typeof call === 'string') {
        return {
            type: 'result',
            result: call,
            status,
            url,
        }
    }
    if (Array.isArray(call)) {
        return {
            type: 'result',
            result: call[1],
            status: call[0],
        }
    }
    if ('ok' in call) {
        return { type: 'response', response: call }
    }
    return responseHandler(call.response, call.status, call.url)
}

/**
 * @param {string} url
 * @param {MaybeResponse | Call} [call]
 * @returns
 */
const getResponse = (url, call) => {
    if (!call) return createResponse(url, {}, INVALID_STATUS)
    const result = responseHandler(call)
    if (result.type === 'response') return result.response
    return createResponse(result.url ?? url, result.result, result.status)
}

/**
 * @type {FetchFn<ExpectedCall | undefined>}
 */
const findNextCall = (input, init) => {
    switch (matchMode) {
        default:
        case MATCH_FIFO:
            return expectedCalls.shift()
        case MATCH_URL: {
            const index = expectedCalls.findIndex(({ match }) => match(input, init))
            if (index === -1) {
                console.error(`Unable to find matching request for ${input} [${init?.body}]`)
                return undefined
            }
            const call = expectedCalls[index]
            expectedCalls = [...expectedCalls.slice(0, index), ...expectedCalls.slice(index + 1)]
            return call
        }
    }
}

/** The heart and soul of this creationâ€¦ */
global.fetch = async (input, init) => {
    const url = input.toString()
    const expectedCall = findNextCall(input, init)
    const response = await new Promise(async (resolve) => {
        if (!expectedCall) console.error(`unexpected fetch request: ${url}`)
        const response = getResponse(url, expectedCall?.response)

        waitingRequests.push({
            url,
            flush: () => resolve(response),
        })
    })

    expect(url).toEqual(response.url)

    calls.push({
        url,
        status: response.status,
        response: response,
        valid: url === response.url,
    })

    return response
}

/**
 * Register a response in the request chain. This will be picked up in order of requests that are made. Don't forget to `act(() => flush(1))` these.
 * @param {any | MaybeResponse} response
 * @param {string | FetchFn<boolean>} url
 * @param {number} [status]
 */
export const on = (response, url = '*', status = 200) => {
    expectedCalls.push({
        match: typeof url === 'string' ? (init) => init.toString() === url : url,
        response:
            typeof response === 'function'
                ? response
                : typeof response === 'object'
                ? JSON.stringify(response)
                : response,
        status,
    })
}

/**
 * Flushes {n} requests. Should be wrapped in `act(() => {â€¦})`-calls. This will allow the react test renderer to keep up with content changes.
 * @param {number} [n]
 */
export const flush = (n = waitingRequests.length) => {
    if (n > waitingRequests.length) throw new Error('Not enough requests to flush')
    while (n--) waitingRequests.pop()?.flush()
}
