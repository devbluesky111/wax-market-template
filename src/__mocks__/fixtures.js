/**
 * @typedef {import("../api/fetch").Auction} Auction
 */

/**
 * @typedef {import("../api/fetch").CollectionData} CollectionData
 */

/**
 * @typedef {import("../api/fetch").Schema} Schema
 */

/**
 * @typedef {import("../api/fetch").Template} Template
 */

/**
 * @typedef {import("../api/fetch").Asset} Asset
 */

/**
 * @param {Partial<CollectionData>} [partial]
 * @returns {CollectionData}
 */
export const collection = (partial = {}) => ({
    allow_notify: false,
    author: 'test-author.wam',
    authorized_accounts: ['test-author.wam'],
    collection_name: 'test-collection',
    contract: 'test-contract',
    created_at_block: '1',
    created_at_time: '1',
    data: {},
    img: 'test-image.svg',
    market_fee: 1001,
    name: 'test-collection',
    notify_accounts: [],
    ...partial,
})

/**
 * @param {Partial<Schema>} [partial]
 * @returns {Schema}
 */
export const schema = (partial = {}) => ({
    contract: 'test-contract',
    created_at_block: '1',
    created_at_time: '1',
    format: [],
    schema_name: 'test-schema',
    ...partial,
    collection: collection(partial.collection),
})

/**
 * @param {Partial<Template>} [partial]
 * @returns {Template}
 */
export const template = (partial = {}) => ({
    contract: 'test-contract',
    created_at_block: '1',
    created_at_time: '1',
    immutable_data: {},
    is_burnable: false,
    is_transferable: false,
    issued_supply: '2',
    max_supply: '3',
    name: 'test-template',
    template_id: 'test-template-id',
    ...partial,
    collection: collection(partial.collection),
    schema: schema(partial.schema),
})

/**
 * @param {Partial<Asset>} [partial]
 * @returns {Asset}
 */
export const asset = (partial = {}) => ({
    asset_id: 'test-id',
    auctions: [],
    backed_tokens: [],
    burned_at_block: '1',
    burned_at_time: '1',
    burned_by_account: 'test.wam',
    contract: 'test-contract',
    data: {},
    immutable_data: null,
    is_burnable: false,
    is_transferable: false,
    minted_at_block: '2',
    minted_at_time: '2',
    mutable_data: null,
    name: 'Test-Asset',
    owner: 'test-owner.wam',
    prices: [],
    sales: [],
    template_mint: '3',
    transferred_at_block: '4',
    transferred_at_time: '4',
    updated_at_block: '5',
    updated_at_time: '5',
    ...partial,
    collection: collection(partial.collection),
    schema: schema(partial.schema),
    template: template(partial.template),
})

/**
 * @param {Partial<Auction>} [partial]
 * @returns {Auction}
 */
export const auction = (partial = {}) => ({
    assets: partial.assets ?? [asset()],
    assets_contract: 'test-asset-contract',
    auction_id: 'test-auction-id',
    bids: [],
    buyer: 'test-buyer',
    claimed_by_buyer: false,
    claimed_by_seller: false,
    collection: collection(partial.collection),
    created_at_block: '1',
    created_at_time: '1',
    end_time: '666',
    is_seller_contract: false,
    maker_marketplace: 'test-maker',
    market_contract: 'test-maker-contract',
    price: {
        amount: '999',
        token_contract: 'test-contract',
        token_precision: 1001,
        token_symbol: 'WAX',
    },
    seller: 'test-seller',
    state: 2,
    taker_marketplace: 'test-marketplace',
    updated_at_block: '2',
    updated_at_time: '2',
})

/**
 * @typedef {import("../api/fetch").NeftyBlend} NeftyBlend
 */

/**
 * @param {Partial<NeftyBlend>} [partial]
 * @returns {NeftyBlend}
 */
export const neftyBlend = (partial = {}) => ({
    blend_id: 'test-blend-id',
    collection_name: 'test-collection',
    display_data: '{}',
    ingredients: [],
    name: 'test-blend',
})

/**
 * @typedef {import("../api/fetch").BlenderizerBlend} BlenderizerBlend
 */

/**
 * @param {Partial<BlenderizerBlend>} [partial]
 * @returns {BlenderizerBlend}
 */
export const blenderizerBlend = (partial = {}) => ({
    collection: 'test-collection',
    mixture: [1, 2, 3, 4, 5],
    owner: 'test-owner',
    target: 1234567890,
})
