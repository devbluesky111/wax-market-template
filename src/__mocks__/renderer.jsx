// ************************************ //
// ************************************ //

// { THIS NEEDS TO BE COPIED CURRENTLY }  //
// This is some quirk with jest, it cannot access this for mocking modules
// as it is itself a â€¦ module :-D

// ************************************ //
// ************************************ //

/** @type {(componentName: string, mapProps?: (props: any) => any) => React.FC<any>} */
export const toJSON =
    (componentName, mapProps = (props) => props) =>
    ({ children, ...props }) =>
        (
            <div>
                <h3>{componentName}</h3>
                <pre>{JSON.stringify(mapProps(props), null, 2)}</pre>
                <div>{children}</div>
            </div>
        )

export const bootstrap = () =>
    jest.mock('react-bootstrap', () => ({
        Tab: toJSON('Tab', ({ title, ...props }) => props),
        Tabs: toJSON('Tabs'),
    }))
