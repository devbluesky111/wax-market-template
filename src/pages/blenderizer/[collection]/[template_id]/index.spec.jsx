import { act, render } from '@testing-library/react'
import { flush, on, reset, setMatchMode } from '../../../../__mocks__/fetch'
import * as fixtures from '../../../../__mocks__/fixtures'
import BlenderizerCollectionTemplateIdPage from './index.page'

jest.mock('../../../../hooks/ual.js')

/** @param {string | BodyInit | undefined | null} data */
// @ts-ignore
const parse = (data) => (data ? JSON.parse(data) : undefined)

describe('::BlenderizerCollectionTemplateIdPage', () => {
    afterEach(reset)

    it('renders correctly', async () => {
        const blend = fixtures.blenderizerBlend()
        const template = fixtures.template()
        on(
            { rows: [blend] },
            (url, init) =>
                url === 'https://api.hivebp.io/v1/chain/get_table_rows' && parse(init?.body).code === 'blenderizerx',
        )
        on(
            {},
            (url, init) =>
                url === 'https://api.hivebp.io/v1/chain/get_table_rows' && parse(init?.body).code === 'marketmapper',
        )
        on(
            { data: template },
            'https://wax.api.atomicassets.io/atomicassets/v1/templates/test-collection/test-template-id',
        )
        on(
            // add the templates we need for the blend
            {
                data: [
                    fixtures.template({ template_id: '1', name: 'A' }),
                    fixtures.template({ template_id: '2', name: 'B' }),
                ],
            },
            'https://wax.api.atomicassets.io/atomicassets/v1/templates?has_assets=true&collection_whitelist=official.wax&limit=1000',
        )

        setMatchMode('url')
        const { container } = render(
            <BlenderizerCollectionTemplateIdPage collectionName={blend.collection} templateId={template.template_id} />,
        )
        expect(container).toMatchSnapshot()
        await act(async () => flush(1))
        expect(container).toMatchSnapshot()
        await act(async () => flush(1))
        expect(container).toMatchSnapshot()
        await act(async () => flush(1))
        expect(container).toMatchSnapshot()
        await act(async () => flush(1))
        expect(container).toMatchSnapshot()
    })
})
