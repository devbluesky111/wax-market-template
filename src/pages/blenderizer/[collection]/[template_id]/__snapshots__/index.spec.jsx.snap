// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`::BlenderizerCollectionTemplateIdPage renders correctly 1`] = `
<div>
  <div
    class="flex justify-center m-auto"
  >
    <img
      alt="Loading"
      class="w-8 h-8 m-auto animate-spin duration-1000"
      data-testid="loading-spinner"
      src="/Loader.svg"
    />
  </div>
</div>
`;

exports[`::BlenderizerCollectionTemplateIdPage renders correctly 2`] = `
<div>
  <div
    class="flex justify-center m-auto"
  >
    <img
      alt="Loading"
      class="w-8 h-8 m-auto animate-spin duration-1000"
      data-testid="loading-spinner"
      src="/Loader.svg"
    />
  </div>
</div>
`;

exports[`::BlenderizerCollectionTemplateIdPage renders correctly 3`] = `
<div>
  <div
    class="w-full h-auto min-h-full overflow-x-hidden m-auto ml-0 z-10 transition transform duration-200"
    id="BlendPage"
  >
    <div
      class="container mx-auto pt-10"
    >
      <div
        class="w-full h-auto text-center md:px-10"
      >
        <div
          class="flex items-center justify-center h-auto"
        >
          <div
            class="w-full md:w-96 flex flex-col items-center"
          >
            <div
              class="relative w-full text-center rounded-md overflow-hidden text-base break-words backdrop-filter backdrop-blur-sm border border-paper shadow-md bg-paper"
            >
              <div
                class="relative flex justify-center asset-img w-full h-auto border p-4 pb-16 border-none"
              >
                <div
                  class="absolute flex justify-evenly w-full bottom-5 t-img-btn"
                />
              </div>
              <div
                class="relative w-full bottom-3 left-1/2 transform -translate-x-1/2"
              >
                test-template
              </div>
            </div>
          </div>
        </div>
        <button
          class="focus:outline-none focus-visible:ring-1 focus-visible:ring-inset focus-visible:ring-primary transition-opacity duration-200 opacity-75 hover:opacity-100 cursor-pointer cursor-not-allowed py-1 px-8 text-secondary mt-3 mb-3 mx-auto cursor-pointer text-sm font-bold leading-relaxed uppercase rounded-3xl outline-none bg-paper"
          disabled=""
        >
          Blend
        </button>
        
        <div
          class="bg-paper px-4 py-2 rounded"
        >
          <div
            class="text-left p-2 text-xl"
          >
            Ingredients
          </div>
          <div
            class="w-full grid grid-cols-4 md:grid-cols-6 2xl:grid-cols-8 gap-2 md:gap-10"
          />
        </div>
        <div
          class="mt-5 bg-paper px-4 py-2 rounded"
        >
          <div
            class="text-left p-2 text-xl"
          >
            My Assets
          </div>
          <div
            class="w-full grid grid-cols-4 md:grid-cols-6 2xl:grid-cols-8 gap-2 md:gap-10"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`::BlenderizerCollectionTemplateIdPage renders correctly 4`] = `
<div>
  <div
    class="w-full h-auto min-h-full overflow-x-hidden m-auto ml-0 z-10 transition transform duration-200"
    id="BlendPage"
  >
    <div
      class="container mx-auto pt-10"
    >
      <div
        class="w-full h-auto text-center md:px-10"
      >
        <div
          class="flex items-center justify-center h-auto"
        >
          <div
            class="w-full md:w-96 flex flex-col items-center"
          >
            <div
              class="relative w-full text-center rounded-md overflow-hidden text-base break-words backdrop-filter backdrop-blur-sm border border-paper shadow-md bg-paper"
            >
              <div
                class="relative flex justify-center asset-img w-full h-auto border p-4 pb-16 border-none"
              >
                <div
                  class="absolute flex justify-evenly w-full bottom-5 t-img-btn"
                />
              </div>
              <div
                class="relative w-full bottom-3 left-1/2 transform -translate-x-1/2"
              >
                test-template
              </div>
            </div>
          </div>
        </div>
        <button
          class="focus:outline-none focus-visible:ring-1 focus-visible:ring-inset focus-visible:ring-primary transition-opacity duration-200 opacity-75 hover:opacity-100 cursor-pointer cursor-not-allowed py-1 px-8 text-secondary mt-3 mb-3 mx-auto cursor-pointer text-sm font-bold leading-relaxed uppercase rounded-3xl outline-none bg-paper"
          disabled=""
        >
          Blend
        </button>
        
        <div
          class="bg-paper px-4 py-2 rounded"
        >
          <div
            class="text-left p-2 text-xl"
          >
            Ingredients
          </div>
          <div
            class="w-full grid grid-cols-4 md:grid-cols-6 2xl:grid-cols-8 gap-2 md:gap-10"
          />
        </div>
        <div
          class="mt-5 bg-paper px-4 py-2 rounded"
        >
          <div
            class="text-left p-2 text-xl"
          >
            My Assets
          </div>
          <div
            class="w-full grid grid-cols-4 md:grid-cols-6 2xl:grid-cols-8 gap-2 md:gap-10"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`::BlenderizerCollectionTemplateIdPage renders correctly 5`] = `
<div>
  <div
    class="w-full h-auto min-h-full overflow-x-hidden m-auto ml-0 z-10 transition transform duration-200"
    id="BlendPage"
  >
    <div
      class="container mx-auto pt-10"
    >
      <div
        class="w-full h-auto text-center md:px-10"
      >
        <div
          class="flex items-center justify-center h-auto"
        >
          <div
            class="w-full md:w-96 flex flex-col items-center"
          >
            <div
              class="relative w-full text-center rounded-md overflow-hidden text-base break-words backdrop-filter backdrop-blur-sm border border-paper shadow-md bg-paper"
            >
              <div
                class="relative flex justify-center asset-img w-full h-auto border p-4 pb-16 border-none"
              >
                <div
                  class="absolute flex justify-evenly w-full bottom-5 t-img-btn"
                />
              </div>
              <div
                class="relative w-full bottom-3 left-1/2 transform -translate-x-1/2"
              >
                test-template
              </div>
            </div>
          </div>
        </div>
        <button
          class="focus:outline-none focus-visible:ring-1 focus-visible:ring-inset focus-visible:ring-primary transition-opacity duration-200 opacity-75 hover:opacity-100 cursor-pointer cursor-not-allowed py-1 px-8 text-secondary mt-3 mb-3 mx-auto cursor-pointer text-sm font-bold leading-relaxed uppercase rounded-3xl outline-none bg-paper"
          disabled=""
        >
          Blend
        </button>
        
        <div
          class="bg-paper px-4 py-2 rounded"
        >
          <div
            class="text-left p-2 text-xl"
          >
            Ingredients
          </div>
          <div
            class="w-full grid grid-cols-4 md:grid-cols-6 2xl:grid-cols-8 gap-2 md:gap-10"
          >
            <div
              class="relative w-full mx-auto rounded-md overflow-hidden flex flex-col text-base break-words backdrop-filter backdrop-blur-sm border transition-all bg-paper"
            >
              <div
                class=""
              >
                <div
                  class="w-full"
                />
                <div>
                  A
                </div>
                <div>
                  Template: 
                  1
                </div>
              </div>
            </div>
            <div
              class="relative w-full mx-auto rounded-md overflow-hidden flex flex-col text-base break-words backdrop-filter backdrop-blur-sm border transition-all bg-paper"
            >
              <div
                class=""
              >
                <div
                  class="w-full"
                />
                <div>
                  B
                </div>
                <div>
                  Template: 
                  2
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mt-5 bg-paper px-4 py-2 rounded"
        >
          <div
            class="text-left p-2 text-xl"
          >
            My Assets
          </div>
          <div
            class="w-full grid grid-cols-4 md:grid-cols-6 2xl:grid-cols-8 gap-2 md:gap-10"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
